Send delete process instance request from gateway to broker

konstantinos-mylosis opened the pull request:
## Description

Delete process instance from secondary storage using Batch Operation providing a single `processInstanceKey`.
If process instance is not present, then throw a `NOT_FOUND`

### Question ❓
`getByKey` method currently requires `READ_PROCESS_INSTANCE` permission. 
Probably if user has `DELETE_PROCESS_INSTANCE` permission, I expect they also have `READ_PROCESS_INSTANCE`, so that shouldn't be a problem.

Do you think we should change this to `anonymous()` for `getByKey` since it is internal logic?

## Related issues

closes #41885

tmetzke commented on zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/batchoperation/BatchOperationCreateProcessor.java:
```diff
@@ -188,6 +188,7 @@ private Either<Rejection, Void> isAuthorized(
             case RESOLVE_INCIDENT -> PermissionType.CREATE_BATCH_OPERATION_RESOLVE_INCIDENT;
             case DELETE_PROCESS_INSTANCE ->
                 PermissionType.CREATE_BATCH_OPERATION_DELETE_PROCESS_INSTANCE;
+            case SINGLE_DELETE_PROCESS_INSTANCE -> PermissionType.DELETE_PROCESS_INSTANCE;
```
💭 The resource type for this permission check will be BATCH (line 196). Is this what we want? A user has to have the permission DELETE_PROCESS_INSTANCE on the resource BATCH to delete a process instance by key? Shouldn't the user have the DELETE_PROCESS_INSTANCE permission on resource PROCESS_DEFINITION, with the resource id being that of the process instance to delete?

konstantinos-mylosis replied:
Based on https://github.com/camunda/camunda/issues/41885
> This request should trigger a batch operation

This was intentional
The idea was to trigger the batch deletion flow even for single deletions so they could run async and reuse the same batch deletion logic.

@remcowesterhoud WDYT?

remcowesterhoud replied:
No, we should not check batch permissions in this case. I should've written this down better in the issue.

The idea is that we use a batch operation for a single process instance deletion. We should make sure if the batch operation type is `SINGLE_DELETE_PROCESS_INSTANCE` we check for the `PROCESS_DEFINITION DELETE_PROCESS_INSTANCE` permission.

@tmetzke you mentioned previously you do the same for single incident resolution. Is this implemented already?

konstantinos-mylosis replied:
💭 Since we do a `getByKey` to check existence and retrieve `ProcessInstanceEntity` in the gateway, can we pass 
process id through `BrokerCreateBatchOperationRequest`?

tmetzke replied:
> @tmetzke you mentioned previously you do the same for single incident resolution. Is this implemented already?

No, not yet, it's tracked here: https://github.com/camunda/camunda/issues/40833

tmetzke replied:
That's how we currently do it, but it's quirky. The issue linked above is meant to correct this.

tmetzke replied:
>**Motivation**
>
>Passing the authorization to check from the gateway to the broker breaks consistency with all other engine commands.

remcowesterhoud replied:
Okay, we're now starting to use this as well 😄 I guess we can replace both once we have a better way.

megglos commented:
> > `getByKey` method currently requires `READ_PROCESS_INSTANCE` permission. Probably if user has `DELETE_PROCESS_INSTANCE` permission, I expect they also have `READ_PROCESS_INSTANCE`, so that shouldn't be a problem.
> > Do you think we should change this to `anonymous()` for `getByKey` since it is internal logic?
> 
> Excellent question! I believe this can go into the architecture discussion even, or be clarified with people like @romansmirnov and @megglos from an architectural and API experience perspective. This is not the only place where we need this, I believe, i.e., if we need to fetch data before we can change data, do we imply a user needs READ and update permissions? We should have a consistent way of applying this.

I haven't checked the details but from high level my understanding is that, this is about a delete functionality on API level only, right? That data is fetched/read as part of performing the delete is an implementation detail, the data is not returned to the user in any way though? In that case I would argue we should not require the READ permission for the read operation given it's internal logic and no data is exposed to the user.

konstantinos-mylosis committed:
feat: create permissions for single process instance deletion

since it is a single process instance deletion, no batch permissions required
create process definition delete process instance authorization

# Conflicts:
#	security/security-core/src/main/java/io/camunda/security/auth/Authorization.java

konstantinos-mylosis committed:
feat: add logic to delete single process instance endpoint

check process instance existence
send broker request to delete process instance
use proper permissions

konstantinos-mylosis committed:
test: single process instance deletion controller

konstantinos-mylosis committed:
test: single process instance deletion service

konstantinos-mylosis commented:
After alignment with @remcowesterhoud 

- `getByKey`: recourceType: `PROCESS_DEFINITION`, permissionType: `DELETE_PROCESS_INSTANCE`, resourceId: `processDefinitionId`
- send `BatchOperationCreationRecord`: recourceType: `PROCESS_DEFINITION`, permissionType: `DELETE_PROCESS_INSTANCE`, resourceId: `processDefinitionId`
- `ProcessInstanceItemProvider`: recourceType: `PROCESS_DEFINITION`, permissionType: `READ_PROCESS_INSTANCE`

konstantinos-mylosis committed:
feat: process instance read requires no permission since it is an internal call

remcowesterhoud reviewed (changes requested):
Nice work @konstantinos-mylosis. Just some smaller things regarding the tests.

konstantinos-mylosis commented on service/src/test/java/io/camunda/service/ProcessInstanceServiceTest.java:
```diff
@@ -527,4 +527,38 @@ final var record = new BatchOperationCreationRecord();
+
+    final var entity = mock(ProcessInstanceEntity.class);
+    when(entity.processInstanceKey()).thenReturn(123L);
+    when(entity.processDefinitionId()).thenReturn("processId");
```
Since Mockito already exists in the test there is no big difference in performance

On the contrary test case using 
- `mock` takes ~5ms
- `Instancio` takes ~450ms

eppdot replied:
I think it's the first call to Instancio that takes a while, because it scans the Class/Object Tree to understand the structure of the the class you wanna instantiate.
